### PR TITLE
Convert WikiSite to kotlin, prelude to Parcelizing it.

### PR DESCRIPTION
**Pull request by dbrant**
https://phabricator.wikimedia.org/T272620

### CODE REVIEW

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.parcelize.Parcelize
+import org.wikipedia.WikipediaApp
+import org.wikipedia.language.AppLanguageLookUpTable
+import org.wikipedia.language.LanguageUtil.firstSelectedChineseVariant
+import org.wikipedia.page.PageTitle
+import org.wikipedia.util.UriUtil.getLanguageVariantFromUri
+import org.wikipedia.util.UriUtil.removeInternalLinkPrefix
+
```
Comment (sharvaniharan) [inline app/src/main/java/org/wikipedia/dataclient/WikiSite.kt:13]: Please remove static imports for readability

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+ * <lh>Validity: authority / language code</lh>
+ *  * Correct: "test.wikipedia.org" / "test"
+ *  * Correct: "wikipedia.org", ""
+ *  * Correct: "no.wikipedia.org", "nb"
+ *  * Incorrect: "wikipedia.org", "test"
+ *
+ */
+@Parcelize
+class WikiSite(
+    @SerializedName("domain") var uri: Uri,
```
Comment (sharvaniharan) [inline app/src/main/java/org/wikipedia/dataclient/WikiSite.kt:40]: Can we please format the parameters next to each other

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+     * @return A [PageTitle] object representing the internalLink passed in.
+     */
+    fun titleForInternalLink(internalLink: String?): PageTitle {
+        // Strip the /wiki/ from the href
+        return PageTitle(removeInternalLinkPrefix(internalLink!!), this)
+    }
+
+    // TODO: this method doesn't have much to do with WikiSite. Move to PageTitle?
+    /**
+     * Create a PageTitle object from a Uri, taking into account any fragment (section title) in the link.
```
Comment (sharvaniharan) [inline app/src/main/java/org/wikipedia/dataclient/WikiSite.kt:141]: This comment could probably be removed?

**Review by sharvaniharan — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt (hunk 3 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/dataclient/WikiSite.kt:141]: What's the problem with it?

Comment (sharvaniharan) [reply app/src/main/java/org/wikipedia/dataclient/WikiSite.kt:137]: We have just been deleting unnecessary comments in favor of self-commenting code during conversion thats all.

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+ *
+ */
+@Parcelize
+class WikiSite(@SerializedName("domain") var uri: Uri, var languageCode: String = "") : Parcelable {
+
+    constructor(uri: Uri) : this(uri, "") {
+        val tempUri = ensureScheme(uri)
+        var authority = tempUri.authority.orEmpty()
+        if (("wikipedia.org" == authority || "www.wikipedia.org" == authority) &&
+            tempUri.path != null && tempUri.path!!.startsWith("/wiki")
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/dataclient/WikiSite.kt]: done!

File: app/src/main/java/org/wikipedia/dataclient/WikiSite.kt [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+        private var DEFAULT_BASE_URL: String? = null
+
+        @JvmStatic
+        fun supportedAuthority(authority: String): Boolean {
+            return authority.endsWith(Uri.parse(DEFAULT_BASE_URL).authority!!)
+        }
+
+        @JvmStatic
+        fun setDefaultBaseUrl(url: String) {
+            DEFAULT_BASE_URL = if (url.isEmpty()) Service.WIKIPEDIA_URL else url
```
[the remark this answers is not in the document]
Comment (dbrant) [reply app/src/main/java/org/wikipedia/dataclient/WikiSite.kt]: done!

**Review by cooltey — APPROVED** [review]
LGTM. @sharvaniharan Please review and merge it if it looks good to you too.

**Comment by sharvaniharan** [comment]
Lgtm... will merge once the conflicts are resolved..

**Comment by dbrant** [comment]
> Lgtm... will merge once the conflicts are resolved..

@sharvaniharan  Ready.

**Pull request merged**